go/libkb/proxy.go

```diff
@@ -95,3 +98,200 @@ func MakeProxy(e *Env) func(r *http.Request) (*url.URL, error) {
+	// auth on the CONNECT request we are sending
+	if s.proxyURL.User != nil {
+		password, _ := s.proxyURL.User.Password()
+		req.SetBasicAuth(s.proxyURL.User.Username(), password)
+	}
+
+	// Send the HTTP request to the proxy server in order to start the TCP tunnel
+	err = req.Write(proxyConn)
+	if err != nil {
+		proxyConn.Close()
+		return nil, err
+	}
+
+	// Read a response and confirm that the server replied with HTTP 200 which confirms that we started the
+	// TCP tunnel. Note that we don't expect any additional body to the request since this is now just an open
+	// TCP tunnel
+	resp, err := http.ReadResponse(bufio.NewReader(proxyConn), req)
+	if err != nil {
+		proxyConn.Close()
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		proxyConn.Close()
+		err = fmt.Errorf("Failed to connect to proxy server, status code: %d", resp.StatusCode)
+		return nil, err
+	}
+
+	// proxyConn is now a TCP connection to the proxy server which forwards to addr. It is the responsibility
+	// of the caller to Close() proxyConn
+	return proxyConn, nil
+}
+
+// Must be called in order for the proxy library to support HTTP connect proxies
+func registerHTTPConnectProxies() {
+	proxy.RegisterDialerType("http", newHTTPConnectProxy)
+	proxy.RegisterDialerType("https", newHTTPConnectProxy)
+}
+
+type ProxyDialOpts struct {
+	Timeout   time.Duration
+	KeepAlive time.Duration
+}
+
+// The equivalent of net.Dial except it uses the proxy configured in Env
+func ProxyDial(env *Env, network string, address string) (net.Conn, error) {
+	// Set the timeout to an exceedingly large number so it never times out
+	return ProxyDialTimeout(env, network, address, 100*365*24*time.Hour)
+}
+
+// The equivalent of net.DialTimeout except it uses the proxy configured in Env
+func ProxyDialTimeout(env *Env, network string, address string, timeout time.Duration) (net.Conn, error) {
+	return ProxyDialWithOpts(context.TODO(), env, network, address, &ProxyDialOpts{Timeout: timeout})
+}
+
+func ProxyDialWithOpts(ctx context.Context, env *Env, network string, address string, opts *ProxyDialOpts) (net.Conn, error) {
+	if env.GetProxyType() == NoProxy {
+		dialer := &net.Dialer{
+			Timeout:   opts.Timeout,
+			KeepAlive: opts.KeepAlive,
+		}
+		return dialer.DialContext(ctx, network, address)
+	}
+	registerHTTPConnectProxies()
+	proxyURLStr := BuildProxyAddressWithProtocol(env.GetProxyType(), env.GetProxy())
+	proxyURL, err := url.Parse(proxyURLStr)
+	if err != nil {
+		return nil, err
+	}
+	dialer, err := proxy.FromURL(proxyURL, getUpstreamDialer(proxyURL, opts))
+	if err != nil {
+		return nil, err
+	}
+
+	// Currently proxy.Dialer does not support DialContext. This is being actively worked on and will probably
+	// land in the next release, but for now we are just dropping the context on the floor
+	// See: https://github.com/golang/go/issues/17759
+	return dialer.Dial(network, address)
```

Comment:
Could we simulate context here with a goroutine that checks for canceled and timeout? That gets used a lot. Something like:
```
doneCh := make(chan net.Conn, 1)
go func() {
  doneCh <- dialer.Dial(network, address)
}()
  select {
    case <-ctx.Done():
        return nil, ctx.Err()
    case ret := <-doneCh:
        return ret, nil
    }
```

Reply:
Ah, clever! I've never seen that pattern before though it definitely works (with the caveat of leaking a goroutine until `dialer.Dial` times out). Tweaked slightly to handle errors but updated.